test/Dialect/FIRRTL/canonicalization.mlir

```diff
@@ -969,6 +969,178 @@ firrtl.module @Reduce(in %a: !firrtl.uint<1>, out %b: !firrtl.uint<1>,
+  %orr2 = firrtl.orr %cat4 : (!firrtl.uint<8>) -> !firrtl.uint<1>
+  firrtl.matchingconnect %out_orr, %orr2 : !firrtl.uint<1>
+
+  // Test XorR with cat containing zero constants - should reduce to xorr(a)
+  // CHECK: %[[XORR_A:.+]] = firrtl.xorr %a
+  // CHECK: firrtl.matchingconnect %out_xorr, %[[XORR_A]]
+  %cat5 = firrtl.cat %c0_ui2, %a : (!firrtl.uint<2>, !firrtl.uint<4>) -> !firrtl.uint<6>
+  %xorr1 = firrtl.xorr %cat5 : (!firrtl.uint<6>) -> !firrtl.uint<1>
+  firrtl.matchingconnect %out_xorr, %xorr1 : !firrtl.uint<1>
+
+  // Test XorR with cat containing non-zero constants - constants should remain
+  // CHECK: %[[CAT_XOR:.+]] = firrtl.cat %c3_ui2, %a
+  // CHECK: %[[XORR_CAT:.+]] = firrtl.xorr %[[CAT_XOR]]
+  // CHECK: firrtl.matchingconnect %out_xorr, %[[XORR_CAT]]
+  %cat6 = firrtl.cat %c3_ui2, %a, %c0_ui2 : (!firrtl.uint<2>, !firrtl.uint<4>, !firrtl.uint<2>) -> !firrtl.uint<8>
+  %xorr2 = firrtl.xorr %cat6 : (!firrtl.uint<8>) -> !firrtl.uint<1>
+  firrtl.matchingconnect %out_xorr, %xorr2 : !firrtl.uint<1>
+}
+
+// CHECK-LABEL: firrtl.module @FlattenCat
+firrtl.module @FlattenCat(in %a: !firrtl.uint<2>, in %b: !firrtl.uint<3>, in %c: !firrtl.uint<4>,
+                         in %d: !firrtl.uint<1>, in %e: !firrtl.uint<2>, in %f: !firrtl.uint<3>,
+                         out %out1: !firrtl.uint<9>, out %out2: !firrtl.uint<15>,
+                         out %out3: !firrtl.uint<10>) {
+  // Test flattening nested cat operations - should flatten into single cat
+  // CHECK: %[[FLAT_CAT:.+]] = firrtl.cat %a, %b, %c
+  // CHECK: firrtl.matchingconnect %out1, %[[FLAT_CAT]]
+  %cat1 = firrtl.cat %a, %b : (!firrtl.uint<2>, !firrtl.uint<3>) -> !firrtl.uint<5>
+  %cat2 = firrtl.cat %cat1, %c : (!firrtl.uint<5>, !firrtl.uint<4>) -> !firrtl.uint<9>
+  firrtl.matchingconnect %out1, %cat2 : !firrtl.uint<9>
+
+  // Test deep nesting (level 4) with multiple operands - should flatten completely
+  // CHECK: %[[DEEP_FLAT:.+]] = firrtl.cat %a, %b, %c, %d, %e, %f
+  // CHECK: firrtl.matchingconnect %out2, %[[DEEP_FLAT]]
+  %inner1 = firrtl.cat %a, %b : (!firrtl.uint<2>, !firrtl.uint<3>) -> !firrtl.uint<5>
+  %inner2 = firrtl.cat %c, %d : (!firrtl.uint<4>, !firrtl.uint<1>) -> !firrtl.uint<5>
+  %middle1 = firrtl.cat %inner1, %inner2 : (!firrtl.uint<5>, !firrtl.uint<5>) -> !firrtl.uint<10>
+  %middle2 = firrtl.cat %e, %f : (!firrtl.uint<2>, !firrtl.uint<3>) -> !firrtl.uint<5>
+  %outer = firrtl.cat %middle1, %middle2 : (!firrtl.uint<10>, !firrtl.uint<5>) -> !firrtl.uint<15>
+  firrtl.matchingconnect %out2, %outer : !firrtl.uint<15>
+
+  // Multiply used cat must not be flatten to prevent exponential blowup
+  // CHECK: %[[AB:.+]] = firrtl.cat %a, %b
+  // CHECK: %[[ABAB:.+]] = firrtl.cat %[[AB]], %[[AB]]
+  // CHECK: firrtl.matchingconnect %out3, %[[ABAB]]
+  %cat_a_b = firrtl.cat %a, %b : (!firrtl.uint<2>, !firrtl.uint<3>) -> !firrtl.uint<5>
+  %cat_a_b_a_b = firrtl.cat %cat_a_b, %cat_a_b : (!firrtl.uint<5>, !firrtl.uint<5>) -> !firrtl.uint<10>
```

Comment:
I think we maybe do want to do this? Did you observe problems in internal designs with this on?

Alternatively, maybe we need a replicate operator for this.

Reply:
Hmm, yeah. I thought it was necessary not to flatten multiply used concat but seems like comb performs flattening for multiply used concat. I'll just remove the condition.